timemachine/ff/amber_converter.py

```diff
@@ -0,0 +1,313 @@
+                        # 2) if we ended up swapping the order (as in the case above), we also need to swap the order of the atoms as well
+                        # 3) record the atom idxs [(atom_6, atom_2), bcc]
+                        src_hash, dst_hash, src_idx, dst_idx = dual_sort(
+                            nb_atom_types_hash[src_type], nb_atom_types_hash[dst_type], src_idx, dst_idx
+                        )
+                        key_hash = src_hash + "|" + dst_hash
+                        if key_hash not in bond_types_to_param_idx:
+                            bond_types_to_param_idx[key_hash] = len(bond_types_to_param_idx)
+
+                        bond_list.append((src_idx, dst_idx))
+                        bond_param_idxs.append(bond_types_to_param_idx[key_hash])
+                        bond_iota += 1
+
+                print("atom symbols", atom_symbols)
+                print("bond list", bond_list)
+                print("bcc parameters", bond_param_idxs)
+
+                # uncomment if you only want to generate types for standard residues
+                # if res.attributes["name"].value.lower() in STANDARD_RESIDUES:
+                all_residues.append(res_name)
+                all_bond_classes.append(bond_param_idxs)
+                res_mol = make_residue_mol(atom_symbols, bond_list)
+                res_mol.SetProp("_Name", res.attributes["name"].value)
+                all_res_mols.append(res_mol)
+                all_bond_idxs.append(bond_list)
+
+    highlightBondLists = []
+    highlightBondColors = []
+
+    # generate colors for bonds
+    for bond_classes in all_bond_classes:
+        bl = []
+        bond_colors_complete = {}
+        for idx in range(len(bond_classes)):
+            bond_colors_complete[idx] = tuple(np.random.random(3))
+        bond_colors = {}
+        for bond_idx, bond_class in enumerate(bond_classes):
+            bl.append(bond_idx)
+            bond_colors[bond_idx] = bond_colors_complete[bond_class]
+
+        highlightBondLists.append(bl)
+        highlightBondColors.append(bond_colors)
+
+    with open("res_out_" + args.method + ".svg", "w") as fh:
+        svg = Draw.MolsToGridImage(
+            all_res_mols,
+            molsPerRow=4,
+            useSVG=True,
+            legends=[x.GetProp("_Name") for x in all_res_mols],
+            highlightBondLists=highlightBondLists,
+            highlightBondColors=highlightBondColors,
+        )
+        fh.write(svg)
+
+    all_patterns_and_params = []
+    for res_name, bond_idxs_list, bond_classes in zip(all_residues, all_bond_idxs, all_bond_classes):
+        bc_dict: Dict[int, List] = {}
+        for bc, bil in zip(bond_classes, bond_idxs_list):
+            if bc not in bc_dict:
+                bc_dict[bc] = []
+            bc_dict[bc].append(bil)
+
+        for v in bc_dict.values():
+            pattern = res_name + " " + str(v)
```

Comment:
nit: Why not make this [res_name, v] ? i.e. a proper nested list?

Reply:
Mostly to generate a "similar looking" list of parameters to our current format. I don't think our current format allows for nested lists, and it current expects a "flat list" of parameters.